File: FizzBuzz/FizzBuzz.cs
using System;

/**
 *
 * Given is the following FizzBuzz application which counts from 1 to 100 and outputs either the corresponding
 * number or the corresponding text if one of the following rules apply.
 * Rules:
 *  - dividable by 3 without a remainder -> Fizz
 *  - dividable by 5 without a remainder -> Buzz
 *  - dividable by 3 and 5 without a remainder -> FizzBuzz
 *
 * ACCEPTANCE CRITERIA:
 * Please refactor this code so that it can be easily extended in the future with other rules, such as
 * "if it is dividable by 7 without a remainder output Bar"
 * "if multiplied by 10 is larger than 100 output Foo"
 *
 */

namespace FizzBuzz
{
    public class Program
    {
        public static void Main(string[] args)
        {
            FizzBuzzEngine engine = new FizzBuzzEngine();

            engine.AddRule(new Rule("Fizz", (n) => { return n % 3 == 0; }));
            engine.AddRule(new Rule("Buzz", (n) => { return n % 5 == 0; }));

            // Example extensions
            engine.AddRule(new Rule("Bar", (n) => { return n % 7 == 0; }));
            engine.AddRule(new Rule("Foo", (n) => { return n * 10 > 100; }));

            engine.Run();
            Console.ReadKey();
        }
    }
}

File: JavaScript/FizzBuzz.js
"use strict";

class Rule {
  constructor(output, condition) {
    if (typeof output !== 'string') {
      console.error('parameter "output" type should be a string')
    }
    this.output = output;
    this.condition = condition;
  }

  Evaluate(operand) {
    if (this.condition(operand)) {
      return this.output;
    }
    return null;
  }
};

class FizzBuzzEngine {
  rules = [];

  AddRule(rule) {
    this.rules.push(rule);
  }

  Run(limit = 100) {
    for (let i = 1; i <= limit; i++) {

      let output = "";

      for (const rule of this.rules) {
        output += rule.Evaluate(i) ?? "";
      }

      if (!output) {
        output = i;
      }

      console.log(`${i}: ${output}`);
    }
  }
}

console.clear();

const engine = new FizzBuzzEngine();

engine.AddRule(new Rule("Fizz", n => n % 3 == 0));
engine.AddRule(new Rule("Buzz", n => n % 5 == 0));

engine.AddRule(new Rule("Bar", n => n % 7 == 0));
engine.AddRule(new Rule("Foo", n => n * 10 > 100));

engine.Run();
